### PR TITLE
Factor out paint recursion into a paint_tree method

### PR DESCRIPTION
**Pull request by chrishtr**
It is a *lot* of changes through various chapters, but I hope things are clearer to the reader. Certainly there is less code
duplication, and likely the `paint_tree` method is easier for readers because it focuses on the recursion, while the
`paint` methods focus only on self-paint and self-paint-effect.

### CODE REVIEW

File: book/layout.md [hunk 1]
```diff
@@ -654,29 +651,32 @@ commands in one place.
 
 ``` {.python}
 class BlockLayout:
-    def paint(self, display_list):
-        for x, y, word, font in self.display_list:
-            display_list.append(DrawText(x, y, word, font))
-        # ...
+    def paint(self):
+        cmds = []
+        if self.layout_mode() == "inline":
```
Comment (pavpanchekha) [inline book/layout.md:656]: Why do we need the test here? If it's not in inline mode, the `display_list` should be empty.

File: book/chrome.md [hunk 2]
```diff
@@ -309,23 +309,25 @@ class TextLayout:
...
 
 ``` {.python}
 class TextLayout:
-    def paint(self, display_list):
+    def paint(self):
+        cmds = []
         color = self.node.style["color"]
-        display_list.append(
+        cmds.append(
             DrawText(self.x, self.y, self.word, self.font, color))
```
Comment (pavpanchekha) [inline book/chrome.md:329]: Probably better to not even construct `cmds` to begin with, just return a one-element list.

File: book/forms.md [hunk 3]
```diff
@@ -171,10 +174,12 @@ Finally, we draw that text:
 
 ``` {.python}
 class InputLayout:
-    def paint(self, display_list):
+    def paint(self):
+        cmds = []
```
Comment (pavpanchekha) [inline book/forms.md:178]: Does this block need `# ...` somewhere?

File: book/forms.md [hunk 4]
```diff
@@ -275,7 +280,7 @@ class BlockLayout:
             if bgcolor != "transparent":
                 x2, y2 = self.x + self.width, self.y + self.height
                 rect = DrawRect(self.x, self.y, x2, y2, bgcolor)
-                display_list.append(rect)
+                cmds.append(rect)
```
Comment (pavpanchekha) [inline book/forms.md:284]: Do we need `# ...` here?

File: book/forms.md [hunk 5]
```diff
@@ -504,11 +509,11 @@ an `input` element is focused:
...
 ``` {.python}
 class InputLayout:
-    def paint(self, display_list):
+    def paint(self):
         # ...
         if self.node.is_focused:
             cx = self.x + self.font.measure(text)
-            display_list.append(DrawLine(
+            cmds.append(DrawLine(
                 cx, self.y, cx, self.y + self.height, "black", 1))
```
Comment (pavpanchekha) [inline book/forms.md:517]: Again a question about `# ...`

File: book/visual-effects.md [hunk 6]
```diff
@@ -803,39 +803,45 @@ class SaveLayer:
...
+display list, we can use `SaveLayer` to add transparency to the whole element.
+I'm going to do this in a new `paint_effects` method, which will wrap `cmds`
+in a `SaveLayer`. The actual `SaveLayer` will be computed in a new
+global `paint_visual_effects` method (because other object types will need it
+also).^[As part of adding this method, I've factored out a `self_rect` and
+`is_atomic` method from `paint`, you'll need to update that method also to call
+these helpers.]
 
 ``` {.python}
 class BlockLayout:
```
Comment (pavpanchekha) [inline book/visual-effects.md:817]: Ok, I'm a bit confused. The reader will need to add `paint_effects` to every single layout class? Do all of them just call `paint_visual_effects`? I think we need to think this through carefully.

File: book/accessibility.md [hunk 7]
```diff
@@ -1355,11 +1351,12 @@ to the browser style sheet above:
...
 ``` {.python}
 class LineLayout:
-    def paint(self, display_list):
+    def paint_effects(self, cmds):
+        # ...
         for child in self.children:
             outline_str = child.node.parent.style.get("outline")
             if parse_outline(outline_str):
-                outline_rect.join(child.rect())
+                outline_rect.join(child.self_rect())
```
Comment (pavpanchekha) [inline book/accessibility.md:1359]: So it looks like we added `self_rect` to every inline object?

File: book/embeds.md [hunk 8]
```diff
@@ -1021,9 +1021,30 @@ Most of the layout tree's `paint` methods don't need to change, but to
 paint an `IframeLayout`, we'll need to paint the child frame:
 
 ``` {.python}
+def paint_tree(layout_object, display_list):
+    cmds = layout_object.paint()
+
+    if isinstance(layout_object, IframeLayout) and \
+        layout_object.node.frame:
+        paint_tree(layout_object.node.frame.document, cmds)
```
Comment (pavpanchekha) [inline book/embeds.md:1029]: Why do this in `paint_tree` instead of in `IframeLayout`?

File: book/embeds.md [hunk 9]
```diff
@@ -1021,9 +1021,30 @@ Most of the layout tree's `paint` methods don't need to change, but to
...
+    cmds = layout_object.paint_effects(cmds)
+    display_list.extend(cmds)
+
+```
+
+Then painting of an `IframeLayout` is just drawing a rectangle and background:
+
+::: {.todo}
+No idea why the expecation doesn't pass.
+:::
```
Comment (pavpanchekha) [inline book/embeds.md:1043]: I can look into it, but sometimes it's mysterious...

File: book/layout.md (hunk 1 above)
Comment (chrishtr) [reply book/layout.md:655]: This code was there before my CL. If it needs to change it should be in a new PR.

File: book/forms.md (hunk 4 above)
Comment (chrishtr) [reply book/forms.md:275]: Not this one.

File: book/accessibility.md (hunk 7 above)
Comment (chrishtr) [reply book/accessibility.md:1361]: Yes. This was the code before?

File: book/embeds.md (hunk 8 above)
Comment (chrishtr) [reply book/embeds.md:1029]: Because I wanted there to be exactly three concepts:

* `paint`, which paints the current layout object
* `paint_effect`, which paints effects applying to the whole subtree (including subframes)
* `paint_tree`, which walks the whole tree across frames and patches together calls to `paint` and `paint_effect` in the right order

File: book/visual-effects.md (hunk 6 above)
Comment (chrishtr) [reply book/visual-effects.md:817]: They don't just blindly call it. For `BlockLayout` it conditions it on being atomic.

**Comment by chrishtr** [comment]
I answered all of the questions, can you re-review?

File: book/visual-effects.md (hunk 6 above)
Comment (pavpanchekha) [reply book/visual-effects.md:817]: I thought about this a lot on the drive home last night and I think it's better to correct for the atomic stuff inside `paint_tree` rather than in `BlockLayout`. For example, suppose you have an `<input>` element with `opacity`. We only want that `opacity` applied once, right? So the atomic inline thing actually applies, I think, to any `BlockLayout` whose root element is not a block element.

Maybe the right move here is to merge this and think about another pass later. But the reason I'm so worried about this PR is that I am worried that we have actual bugs here, and like usual I'm trying to achieve "obviously no bugs" and not yet succeeding.

File: book/accessibility.md (hunk 7 above)
Comment (pavpanchekha) [reply book/accessibility.md:1361]: I think it's fine for the PR, but I will say that I'm becoming more and more convinced that we need to add a `Rect` class in Chapter 7 and then set up the expectation of each layout object having a `rect` method earlier.

File: book/embeds.md (hunk 8 above)
Comment (pavpanchekha) [reply book/embeds.md:1029]: I guess that's fine. I don't have a strong theory of where exactly per-layout-object and paint-generic operations go.

**Review by pavpanchekha — APPROVED** [review]

**Comment by pavpanchekha** [comment]
Ok, I'm approving this for merge. I continue to worry about atomic inlines—in fact, maybe we have bugs here—but the changes are clearly a net positive for the book.

File: book/accessibility.md (hunk 7 above)
Comment (chrishtr) [reply book/accessibility.md:1361]: We did agree last week to add the rect class. :)

File: book/visual-effects.md (hunk 6 above)
Comment (chrishtr) [reply book/visual-effects.md:817]: Ok I'll land it now and we can test & make another PR to try out moving is_atomic.

p.s. not sure what you mean by "whose root element is not a block element". Are you referring to it being for an inline element with some blocks inside it?
